### ISSUE TITLE
Request: offer a new function to check how much the dynamic-colors are supported on the device, and not just a boolean

### ISSUE THREAD

**Issue by AndroidDeveloperLB** [material-components/material-components-android#2779]
**Is your feature request related to a problem? Please describe.**
As it seems, isDynamicColorAvailable isn't reliable and just returns a boolean instead of telling us how much the device supports dynamic colors:
https://github.com/material-components/material-components-android/issues/2775

I was told that OEMs can still do whatever they want with the colors, including not doing anything to the OS and other apps:
https://github.com/material-components/material-components-android/issues/2775#issuecomment-1165681852

**Describe the solution you'd like**
The function would be a part of the OS framework itself (and of the support library if possible), and won't return a boolean.
Instead it will return an EnumSet (or a combined flags, of course) of these:

1. Affects OS colors
2. Affects apps colors

If none of these are set, it means it's not supported at all (because what's the point in having this if it doesn't affect anything?).

**Describe alternatives you've considered**
I could buy all the devices out there, but I think it's better that Google will handle this via an official API that OEMs must follow.

**Additional context**
Also requested here:
https://issuetracker.google.com/issues/234050622#comment6

**Comment by drchen** [comment]
Android S doesn't have such API implemented and since Android T all OEMs will have to support dynamic colors.

I'll close this issue as infeasible (on S) and outdated (on T). Also it's something beyond the Material library's control.

**Comment by AndroidDeveloperLB** [comment]
@drchen You haven't read what I wrote.
I was told that OEMs won't have to use the colors, not for the OS and not for other apps.
Because of this, the feature might say it's supported, yet it won't affect anything or won't be noticeable in various places.
The request is that it will tell us what it does support. Not that it's "just supports".

Please re-read what I wrote. I wrote that I don't want a boolean. I want an EnumSet.

**Comment by drchen** [comment]
We implemented `DynamicColors.isDynamicColorAvailable()` according to whatever info we have from the partners. 

I don't think OEM can only support dynamic colors partially. (I mean, of course they can do that if they want to, but I have no idea if Google allows them to do that. I guess probably not.)

And as I said, since T, all OEMs will have to support dynamic colors, so this won't be an issue anymore on T.

And as for S - according to my personal understanding (I'm *not* a product person so I *cannot* speak for Google as for this part) I don't think any OEMs are required to support dynamic colors. The only thing we can do from the library side is using our best efforts to update the list when our product team let us know when an OEM starts supporting dynamic colors.

Again, no matter it's a boolean or an enum, unfortunately what we have implemented right now is already the best way to tell if an OEM device supports dynamic colors on S. It's a limitation on the framework side and we are already using the best info from the product as much as we can.

**Comment by AndroidDeveloperLB** [comment]
@drchen 
1. So you need to talk more with the partners and also see that it works as you'd expect.

2. As for "doing what they want", this is what I got from the comment:
" It's then up to the OS and other apps such as Gmail to decide whether they are using those colors."

3. I already know that Gmail and various Google apps should change their colors. Why don't they, if it says it's supported?

4.What is your definition of " tell if an OEM device supports dynamic colors on S" , if it doesn't affect any app, not even the OS itself? 

5. Why isn't there an option in the settings to choose the colors? 

6. Would Android T support it AND update its UI by default, depending on the colors? 
Or again says it has support, and still OEMs can do whatever they want with it, including nothing at all and including not offering it to apps?

**Comment by drchen** [comment]
> So you need to talk more with the partners and also see that it works as you'd expect.

As an engineer, what I can do is reporting it to the product team when we see some OEMs are not working as expected. 

> As for "doing what they want", this is what I got from the comment:
> " It's then up to the OS and other apps such as Gmail to decide whether they are using those colors."

Yes. This description is accurate. So:

1. Apps can decide if they want to opt in dynamic color support. Android framework has no control over it.
2. On S, it's up to OEM if they want to support dynamic colors and to what extent (as far as I know)
 
> I already know that Gmail and various Google apps should change their colors. Why don't they, if it says it's supported?

Because the OEM OS may not support dynamic colors. 
 
> What is your definition of " tell if an OEM device supports dynamic colors on S" , if it doesn't affect any app, not even the OS itself?

So this is how dynamic colors work:

1. Android framework defined a series of system color resources, which are supposed to be overridden at runtime according to the wallpaper color users selected.
2. Apps use those color resources directly or indirectly (say, via the Material library) in their UI. 
3. System Apps are also apps, so they also need to use those color resources explicitly to have dynamic color support.

Some OEM OS may not be doing step 1 & 3 or in a different way on S. Step 2 & 3 are totally optional and up to each app's implementation. We have no control over them. 

As a library, we have no technical way to know how an OEM OS is implementing Step 1, not mentioning Step 3. 

> Why isn't there an option in the settings to choose the colors?

I have no idea. That's a framework or OEM decision. 

> Would Android T support it AND update its UI by default, depending on the colors?
> Or again says it has support, and still OEMs can do whatever they want with it, including nothing at all and including not offering it to apps?

As far as I know they should. But I'm not a product person so I cannot speak for them. Also the Material library is not part of the Android framework. It will be easier to get the answer you need by reaching out to the framework team, I think. : )

**Comment by AndroidDeveloperLB** [comment]
1. Have you tested on various devices out there, of various OEMs, and confirmed it works properly?
2. So again, how would I test that dynamic colors work or not, if OEMs can do whatever they want yet declare they support it? I tried to show you on the sample app, but you say that they can just avoid generating different colors and provide the same ones all the time.
3. And as I wrote, the library says it does support. 
4. If not by the library, will it be possible on Android 13 to detect if the OS truly supports dynamic colors, meaning it should affect apps that use it, and also the OS itself (notification drawer, volume control, ...) ? Or it will still be fragmented and unknown?
5. It seems like a bug, because what's the point in having a horizontal list of colors with just one option...
6. I already requested this on the issue tracker. Please tell them to check it out

**Comment by drchen** [comment]
I agree it's not perfect. But again it's the best we can do on the library side and it's better than nothing.

Please be aware that the role of the Material library plays in dynamic color support is a "helper". We tried to make it easier for clients to integrate dynamic colors in their apps. Yes unfortunately we are not able to cover all OEM OS versions (it's not possible for us to audit all of them either), but I think it's better than nothing. (The worse case is users won't see dynamic colors on the apps that are supposed to support them but on those OS there's no dynamic color support anyway.)

The framework team may have their concerns which I don't really know. So reaching out to them will be the best way to raise your points. I will let them know our user feedback but that's the best I can do.

**Comment by AndroidDeveloperLB** [comment]
@drchen Thank you.
Can you please also contact the Android Studio team, so that new projects created on the IDE will offer a checkbox (turned on by default) to use the dynamic colors?
So far none of the templates there use what's used here.
This can help to encourage developers to use it.

The sample here is nice, but it's way more complex for developers to start with it. It has many modules in it and hard to navigate and see what is the minimal thing to do for each example...
